netket/graph/common_lattices.py

```diff
@@ -0,0 +1,248 @@
+    obc = np.logical_not(pbc)
+    ndim = len(extent)
+    for perm in permutations(axes):
+        if np.all(extent == extent[list(perm)]) and np.all(
+            axes[obc] == axes[list(perm)][obc]
+        ):
+            axis_perm.append(_perm_symm(perm))
+    result = PointGroup(axis_perm, ndim=ndim)
+    # reflections across axes
+    # can only do it across periodic axes
+    for i in axes[pbc]:
+        result = result @ PointGroup([Identity(), _axis_reflection(i, ndim)], ndim=ndim)
+    result = result.elems
+    result[0] = Identity()  # it would otherwise be an equivalent PGSymmetry
+    return PointGroup(result, ndim=ndim)
+    return result
+
+
+def Grid(length: Sequence[int], *, pbc: Union[bool, Sequence[bool]] = True) -> Lattice:
+    """
+    Constructs a hypercubic lattice given its extent in all dimensions.
+
+    Args:
+        length: Side length of the lattice. It must be a list with integer components >= 1.
+        pbc: If `True`, the grid will have periodic boundary conditions (PBC);
+             if `False`, the grid will have open boundary conditions (OBC).
+             This parameter can also be a list of booleans with same length as
+             the parameter `length`, in which case each dimension will have
+             PBC/OBC depending on the corresponding entry of `pbc`.
+
+    Examples:
+        A 5x10 lattice with periodic boundary conditions can be
+        constructed as follows:
+
+        >>> import netket
+        >>> g=netket.graph.Grid(length=[5, 10], pbc=True)
+        >>> print(g.n_nodes)
+        50
+
+        Also, a 2x2x3 lattice with open boundary conditions can be constructed as follows:
+
+        >>> g=netket.graph.Grid(length=[2,2,3], pbc=False)
+        >>> print(g.n_nodes)
+        12
+    """
+    length = np.asarray(length, dtype=int)
+    ndim = len(length)
+    if isinstance(pbc, bool):
+        pbc = [pbc] * ndim
+    return Lattice(
+        basis_vectors=np.eye(ndim),
+        extent=length,
+        pbc=pbc,
+        point_group=_grid_point_group(length, pbc),
+    )
+
+
+def Hypercube(length: int, n_dim: int = 1, *, pbc: bool = True) -> Lattice:
+    r"""Constructs a hypercubic lattice with equal side length in all dimensions.
+    Periodic boundary conditions can also be imposed.
+
+    Args:
+        length: Side length of the hypercube; must always be >=1
+        n_dim: Dimension of the hypercube; must be at least 1.
+        pbc: Whether the hypercube should have periodic boundary conditions (in all directions)
+
+    Examples:
+         A 10x10x10 cubic lattice with periodic boundary conditions can be
+         constructed as follows:
+
+         >>> import netket
+         >>> g = netket.graph.Hypercube(10, n_dim=3, pbc=True)
+         >>> print(g.n_nodes)
+         1000
+    """
+    length_vector = [length] * n_dim
+    return Grid(length_vector, pbc=pbc)
+
+
+Square = partial(Hypercube, n_dim=2)
+r"""Constructs a square lattice of side `length`
```

Comment:
Does this work if you do `help(nk.graph.Square)`? it gives the right docstring?

If not, you might need to do `Square.__doc__ = square_docstring`

Reply:
Oops

Reply:
EDIT it works in the sense that `Square.__doc__` returns the correct docstring, but `help(nk.graph.Square)` produces a generic help string about `partial`

Reply:
There is [`functools.update_wrapper`](https://docs.python.org/3/library/functools.html#functools.update_wrapper), which transfers metadata more thoroughly and might help here.

Though just defining
```python
def Square(...):
    """ ... """
    return Hypercube(..., n_dim=2)
```
is probably the easiest solution.

Reply:
i agree

Reply:
yaaay